src/modules/musig/main_impl.h

```diff
@@ -150,286 +160,302 @@ int secp256k1_musig_pubkey_tweak_add(const secp256k1_context* ctx, secp256k1_mus
+            }
+            ret = secp256k1_pubkey_load(ctx, &noncep, &nonce);
+            /* Successfully parsed pubkey is always valid */
+            VERIFY_CHECK(ret);
+            secp256k1_gej_add_ge_var(&summed_nonces[j], &summed_nonces[j], &noncep, NULL);
+        }
     }
-    (void) ctx;
+}
 
-    memset(session, 0, sizeof(*session));
 
-    session->magic = session_magic;
-    memcpy(&session->combined_pk, combined_pk, sizeof(*combined_pk));
-    session->pre_session = *pre_session;
-    session->n_signers = (uint32_t) n_signers;
-    secp256k1_musig_signers_init(signers, session->n_signers);
+int secp256k1_musig_nonces_combine(const secp256k1_context* ctx, unsigned char *combined_pubnonce66, const unsigned char * const* pubnonces, size_t n_pubnonces) {
+    secp256k1_gej summed_nonces[2];
+    int i;
 
-    session->pre_session = *pre_session;
-    session->is_msg_set = 1;
-    memcpy(session->msg, msg32, 32);
-    session->has_secret_data = 0;
+    ARG_CHECK(combined_pubnonce66 != NULL);
+    ARG_CHECK(pubnonces != NULL);
+    ARG_CHECK(n_pubnonces > 0);
 
-    for (i = 0; i < n_signers; i++) {
-        memcpy(signers[i].nonce_commitment, commitments[i], 32);
+    secp256k1_musig_sum_nonces(ctx, summed_nonces, pubnonces, n_pubnonces);
+    for (i = 0; i < 2; i++) {
+        int ret;
+        secp256k1_ge noncep;
+        size_t len = 33;
+        if (secp256k1_gej_is_infinity(&summed_nonces[i])) {
+            return 0;
+        }
+        secp256k1_ge_set_gej(&noncep, &summed_nonces[i]);
+        ret = secp256k1_eckey_pubkey_serialize(&noncep, &combined_pubnonce66[i*33], &len, 1);
+        VERIFY_CHECK(ret);
     }
-    session->round = 1;
     return 1;
 }
 
-int secp256k1_musig_set_nonce(const secp256k1_context* ctx, secp256k1_musig_session_signer_data *signer, const unsigned char *nonce) {
-    secp256k1_sha256 sha;
-    unsigned char commit[32];
-
-    VERIFY_CHECK(ctx != NULL);
-    ARG_CHECK(signer != NULL);
-    ARG_CHECK(nonce != NULL);
+static void secp256k1_musig_session_cache_load(secp256k1_scalar *b, secp256k1_scalar *e, int *combined_nonce_parity, const secp256k1_musig_session_cache *cache) {
+    secp256k1_scalar_set_b32(b, &cache->data[0], NULL);
+    secp256k1_scalar_set_b32(e, &cache->data[32], NULL);
+    *combined_nonce_parity = cache->data[64];
+}
 
-    secp256k1_sha256_initialize(&sha);
-    secp256k1_sha256_write(&sha, nonce, 32);
-    secp256k1_sha256_finalize(&sha, commit);
+static void secp256k1_musig_template_load(unsigned char *r32, secp256k1_scalar *s, const secp256k1_musig_template *template) {
+    memcpy(r32, &template->data[0], 32);
+    secp256k1_scalar_set_b32(s, &template->data[32], NULL);
+}
 
-    if (memcmp(commit, signer->nonce_commitment, 32) != 0) {
-        return 0;
-    }
-    memcpy(&signer->nonce, nonce, sizeof(*nonce));
-    if (!secp256k1_xonly_pubkey_parse(ctx, &signer->nonce, nonce)) {
+static int secp256k1_xonly_ge_serialize(unsigned char *output32, secp256k1_ge *ge) {
```

Comment:
this function has a side effect on `ge->x` since it calls `secp256k1_fe_normalize_var(&ge->x);`

Reading the name of the function, this is a surprising fact. I would have assumed serialize does not change any internal state.

Reply:
This function is not exposed in the public API. It's pretty common for functions internal to libsecp to modify their arguments by normalizing them, and this is usually not too hard to notice (for people working frequently on the library) because it interferes with const-correctness (which I guess is how you noticed this).

Reply:
@apoelstra yes, I wanted to make sure, since I saw places where const was forgotten (fixed by now) + the fact the method name did not suggested that, wanted to be sure it is not a mistake.

Reply:
Added comment explaining this fact.